File: MCRunner/Instruments/MonitoredBars.cs
using System.Collections.Generic;
using System;

namespace MCRunner.Instruments
{
    public class MonitoredBars : LoadableBars, IMonitoredInstrument
    {
        public event Action<Bar> Updated;

        public override void AddBar(Bar bar)
        {
            base.AddBar(bar);
            Updated.SafeTrigger(bar);
        }

        public new IEnumerable<IMonitoredInstrument> ToSingleDataStream()
        {
            return new List<IMonitoredInstrument> { this };
        }
    }
}

File: python/mcrunner/instruments/monitored_bars.py
from typing import Callable, List
from .loadable_bars import LoadableBars
from .bar import Bar

class MonitoredBars(LoadableBars):
    def __init__(self):
        super().__init__()
        self._callbacks: List[Callable[[Bar], None]] = []

    def add_callback(self, cb: Callable[[Bar], None]):
        self._callbacks.append(cb)

    def add_bar(self, bar: Bar):
        super().add_bar(bar)
        for cb in list(self._callbacks):
            cb(bar)

    def to_single_data_stream(self):
        return [self]
